pkg/messaging/topics.go

```diff
@@ -0,0 +1,156 @@
+	}
+
+	return domainID, chanID, subtopic, nil
+}
+
+func ParsePublishSubtopic(subtopic string) (parseSubTopic string, err error) {
+	if subtopic == "" {
+		return subtopic, nil
+	}
+
+	subtopic, err = formatSubtopic(subtopic)
+	if err != nil {
+		return "", errors.Wrap(ErrMalformedSubtopic, err)
+	}
+
+	if strings.ContainsAny(subtopic, subtopicInvalidChars+wildcards) {
+		return "", ErrMalformedSubtopic
+	}
+
+	if strings.Contains(subtopic, "..") {
+		return "", ErrMalformedSubtopic
+	}
+
+	return subtopic, nil
+}
+
+func ParseSubscribeTopic(topic string) (domainID string, chanID string, subtopic string, err error) {
+	msgParts := topicRegExp.FindStringSubmatch(topic)
+	if len(msgParts) < numGroups {
+		return "", "", "", ErrMalformedTopic
+	}
+
+	domainID = msgParts[domainGroup]
+	chanID = msgParts[channelGroup]
+	subtopic = msgParts[subtopicGroup]
+	subtopic, err = ParseSubscribeSubtopic(subtopic)
+	if err != nil {
+		return "", "", "", errors.Wrap(ErrMalformedTopic, err)
+	}
+
+	return domainID, chanID, subtopic, nil
+}
+
+func ParseSubscribeSubtopic(subtopic string) (parseSubTopic string, err error) {
+	if subtopic == "" {
+		return "", nil
+	}
+
+	if strings.ContainsAny(subtopic, mqWildcards) {
+		subtopic = wildcardsReplacer.Replace(subtopic)
+	}
+	subtopic, err = formatSubtopic(subtopic)
+	if err != nil {
+		return "", errors.Wrap(ErrMalformedSubtopic, err)
+	}
+
+	if strings.ContainsAny(subtopic, subtopicInvalidChars) {
+		return "", ErrMalformedSubtopic
+	}
+
+	if strings.Contains(subtopic, "..") {
+		return "", ErrMalformedSubtopic
+	}
+
+	for _, elem := range strings.Split(subtopic, ".") {
+		if len(elem) > 1 && strings.ContainsAny(elem, wildcards) {
+			return "", ErrMalformedSubtopic
+		}
+	}
+	return subtopic, nil
+}
+
+func formatSubtopic(subtopic string) (string, error) {
+	subtopic, err := url.QueryUnescape(subtopic)
+	if err != nil {
+		return "", err
+	}
+	subtopic = strings.TrimPrefix(subtopic, "/")
+	subtopic = strings.TrimSuffix(subtopic, "/")
+	subtopic = strings.TrimSpace(subtopic)
+	subtopic = pathReplacer.Replace(subtopic)
+	return subtopic, nil
+}
+
+func EncodeTopic(domainID string, channelID string, subtopic string) string {
+	return fmt.Sprintf("%s.%s", MsgTopicPrefix, EncodeTopicSuffix(domainID, channelID, subtopic))
+}
+
+func EncodeTopicSuffix(domainID string, channelID string, subtopic string) string {
+	subject := fmt.Sprintf("%s.%s.%s", domainID, ChannelTopicPrefix, channelID)
+	if subtopic != "" {
+		subject = fmt.Sprintf("%s.%s", subject, subtopic)
+	}
+	return subject
+}
+
+func (m *Message) EncodeTopicSuffix() string {
+	return EncodeTopicSuffix(m.GetDomain(), m.GetChannel(), m.GetSubtopic())
+}
+
+func (m *Message) EncodeToMQTTTopic() string {
+	topic := fmt.Sprintf("%s/%s/%s/%s", MsgTopicPrefix, m.GetDomain(), ChannelTopicPrefix, m.GetChannel())
+	if m.GetSubtopic() != "" {
+		topic = topic + "/" + strings.ReplaceAll(m.GetSubtopic(), ".", "/")
+	}
+	return topic
+}
```

Comment:
Can we not attach methods to the generated code? I am afraid it is using reflection under the hood, and this may mess up with Protobuf. Probably will not, but there is no need to attach it to the message anyway.